modules/artifact/src/main/java11/org/mule/runtime/module/artifact/api/classloader/FineGrainedControlClassLoader.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.artifact.api.classloader;
+
+import static java.lang.Boolean.valueOf;
+import static java.lang.String.format;
+import static java.lang.System.getProperty;
+import static org.mule.runtime.api.util.MuleSystemProperties.MULE_LOG_VERBOSE_CLASSLOADING;
+import static org.mule.runtime.api.util.Preconditions.checkArgument;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.mule.api.annotation.NoInstantiate;
+import org.mule.runtime.core.api.util.ClassUtils;
+import org.mule.runtime.core.api.util.CompoundEnumeration;
+import org.mule.runtime.module.artifact.api.classloader.exception.CompositeClassNotFoundException;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+
+import org.slf4j.Logger;
+
+/**
+ * Defines a {@link ClassLoader} which enables the control of the class loading lookup mode.
+ * <p/>
+ * By using a {@link ClassLoaderLookupPolicy} this classLoader can use parent-first, parent-only or child-first classloading
+ * lookup mode per package.
+ */
+@NoInstantiate
+public class FineGrainedControlClassLoader extends URLClassLoader
```

Comment:
this is a very big class and with a lot of duplicate code to put in an MRJAr, please make the variable parts of the mrjar as compact as possible (and putthing them in thei own module should be clearer, also)